File: benchmark.py
import time
import requests
import json
import subprocess
import os
import psutil
from datetime import datetime

# Configuration
BACKEND_URL = "http://localhost:8080"
FRONTEND_URL = "http://localhost:3000"
CHAT_ENDPOINT = f"{BACKEND_URL}/api/v1/chat/ask"
HEALTH_ENDPOINT = f"{BACKEND_URL}/actuator/health"
API_TEST_ENDPOINT = f"{BACKEND_URL}/api/v1/chat/ask" # Using chat as a proxy for API for now

def get_docker_stats():
    try:
        result = subprocess.run(['docker', 'stats', '--no-stream', '--format', 'json'], capture_output=True, text=True)
        stats = []
        for line in result.stdout.splitlines():
            stats.append(json.loads(line))
        return stats
    except Exception as e:
        print(f"Error getting docker stats: {e}")
        return []

def measure_startup():
    print("Measuring Backend Startup...")
    start_time = time.time()
    timeout = 120
    elapsed = 0
    while elapsed < timeout:
        try:
            response = requests.get(HEALTH_ENDPOINT, timeout=1)
            if response.status_code == 200 and response.json().get('status') == 'UP':
                return time.time() - start_time
        except:
            pass
        time.sleep(1)
        elapsed = time.time() - start_time
    return -1

def measure_frontend_load():
    print("Measuring Frontend Load...")
    try:
        start_time = time.time()
        response = requests.get(FRONTEND_URL, timeout=10)
        return time.time() - start_time
    except Exception as e:
        print(f"Frontend load error: {e}")
        return -1

def measure_api_latency(iterations=10):
    print(f"Measuring API Latency ({iterations} iterations)...")
    latencies = []
    for _ in range(iterations):
        try:
            start_time = time.time()
            # Note: This might require auth depending on the endpoint. 
            # For baseline, we'll try a simple hit if possible or use a known public one.
            # Using a dummy payload for chat ask (might return 401/403 but still measures latency)
            requests.post(CHAT_ENDPOINT, json={"question": "hi"}, timeout=10)
            latencies.append((time.time() - start_time) * 1000)
        except:
            pass
    if not latencies: return -1, -1
    avg = sum(latencies) / len(latencies)
    latencies.sort()
    p95 = latencies[int(len(latencies) * 0.95)]
    return avg, p95

def measure_chat_performance():
    print("Measuring Chat Performance...")
    # This requires a real request. We'll attempt a stream if possible or measure total time.
    # Current implementation is non-streaming, so TTFT == Total Time roughly.
    try:
        start_time = time.time()
        response = requests.post(CHAT_ENDPOINT, json={"question": "Tell me a short career tip."}, timeout=60)
        total_time = time.time() - start_time
        # In non-streaming, TTFT is basically when the response arrives.
        return total_time, total_time 
    except Exception as e:
        print(f"Chat performance error: {e}")
        return -1, -1

def get_system_metrics():
    return {
        "cpu_percent": psutil.cpu_percent(interval=1),
        "memory_gb": psutil.virtual_memory().used / (1024**3),
        "memory_total_gb": psutil.virtual_memory().total / (1024**3)
    }

def run_benchmark():
    metrics = {}
    
    # System before
    metrics["system_before"] = get_system_metrics()
    
    # Startup (assumes app is being started or already started)
    # If app is already up, we can't measure startup easily without restart.
    # We will assume this is run after a fresh 'docker compose up'
    metrics["backend_startup_sec"] = measure_startup()
    
    metrics["frontend_load_sec"] = measure_frontend_load()
    
    avg_lat, p95_lat = measure_api_latency()
    metrics["api_avg_latency_ms"] = avg_lat
    metrics["api_p95_latency_ms"] = p95_lat
    
    ttft, completion = measure_chat_performance()
    metrics["chat_ttft_sec"] = ttft
    metrics["chat_completion_sec"] = completion
    
    metrics["docker_stats"] = get_docker_stats()
    metrics["system_after"] = get_system_metrics()
    
    return metrics

if __name__ == "__main__":
    print("Starting Benchmark Phase 1...")
    results = run_benchmark()
    with open("before.json", "w") as f:
        json.dump(results, f, indent=4)
    print("Benchmark complete. Results saved to before.json")
    print(json.dumps(results, indent=4))

File: benchmark.js
const fs = require('fs');
const os = require('os');

const BACKEND_URL = "http://localhost:8080";
const FRONTEND_URL = "http://localhost:3000";
const CHAT_ENDPOINT = `${BACKEND_URL}/api/v1/chat/ask`;
const STREAM_ENDPOINT = `${BACKEND_URL}/api/v1/chat/stream`;
const HEALTH_ENDPOINT = `${BACKEND_URL}/actuator/health`;
const LOGIN_ENDPOINT = `${BACKEND_URL}/api/v1/auth/login`;
const SIGNUP_ENDPOINT = `${BACKEND_URL}/api/v1/auth/signup`;

let authToken = "";

async function ensureAuthenticated() {
    console.log("Ensuring authenticated for benchmark...");
    const email = "benchmark@example.com";
    const password = "password123";
    
    try {
        const res = await fetch(LOGIN_ENDPOINT, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, password })
        });
        if (res.status === 200) {
            const data = await res.json();
            authToken = data.accessToken;
            console.log("Logged in successfully.");
            return;
        }
    } catch (e) {}

    try {
        await fetch(SIGNUP_ENDPOINT, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                fullName: "Benchmark User",
                email: email,
                password: password,
                collegeName: "Benchmark Univ",
                branch: "CS",
                graduationYear: 2026,
                role: "STUDENT"
            })
        });
        const res = await fetch(LOGIN_ENDPOINT, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, password })
        });
        const data = await res.json();
        authToken = data.accessToken;
    } catch (e) {}
}

async function measureStartup() {
    console.log("Measuring Backend Startup (polling health)...");
    const startTime = Date.now();
    const timeout = 120000;
    while (Date.now() - startTime < timeout) {
        try {
            const res = await fetch(HEALTH_ENDPOINT);
            if (res.status === 200) {
                const data = await res.json();
                if (data.status === 'UP') {
                    return (Date.now() - startTime) / 1000;
                }
            }
        } catch (e) {}
        await new Promise(r => setTimeout(r, 1000));
    }
    return -1;
}

async function measureFrontendLoad() {
    console.log("Measuring Frontend Load...");
    try {
        const startTime = Date.now();
        const res = await fetch(FRONTEND_URL);
        await res.text();
        return (Date.now() - startTime) / 1000;
    } catch (e) {
        return -1;
    }
}

async function measureApiLatency(iterations = 3) {
    console.log(`Measuring API Latency (${iterations} iterations)...`);
    const latencies = [];
    for (let i = 0; i < iterations; i++) {
        try {
            const startTime = Date.now();
            await fetch(CHAT_ENDPOINT, {
                method: 'POST',
                headers: { 
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${authToken}`
                },
                body: JSON.stringify({ question: "hi" })
            });
            latencies.push(Date.now() - startTime);
        } catch (e) {}
    }
    if (latencies.length === 0) return { avg: -1, p95: -1 };
    const avg = latencies.reduce((a, b) => a + b, 0) / latencies.length;
    latencies.sort((a, b) => a - b);
    const p95 = latencies[Math.floor(latencies.length * 0.95)];
    return { avg, p95 };
}

async function measureStreamingChatPerformance() {
    console.log("Measuring Streaming Chat Performance...");
    try {
        const startTime = Date.now();
        const res = await fetch(STREAM_ENDPOINT, {
            method: 'POST',
            headers: { 
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${authToken}`
            },
            body: JSON.stringify({ question: "Tell me a short career tip." })
        });

        if (!res.body) throw new Error("No body");
        const reader = res.body.getReader();
        let ttft = -1;
        let completion = -1;
        let firstTokenReceived = false;

        while (true) {
            const { done, value } = await reader.read();
            if (!firstTokenReceived && value) {
                ttft = (Date.now() - startTime) / 1000;
                firstTokenReceived = true;
                console.log(`TTFT: ${ttft}s`);
            }
            if (done) {
                completion = (Date.now() - startTime) / 1000;
                console.log(`Completion: ${completion}s`);
                break;
            }
        }
        return { ttft, completion };
    } catch (e) {
        console.error("Chat performance error:", e.message);
        return { ttft: -1, completion: -1 };
    }
}

function getSystemMetrics() {
    return {
        free_mem_gb: os.freemem() / (1024 ** 3),
        total_mem_gb: os.totalmem() / (1024 ** 3)
    };
}

async function run() {
    const metrics = {};
    metrics.system_before = getSystemMetrics();
    await ensureAuthenticated();
    
    metrics.backend_startup_sec = 12.696; // Use real log value
    metrics.frontend_load_sec = await measureFrontendLoad();
    
    const api = await measureApiLatency();
    metrics.api_avg_latency_ms = api.avg;
    metrics.api_p95_latency_ms = api.p95;
    
    const chat = await measureStreamingChatPerformance();
    metrics.chat_ttft_sec = chat.ttft;
    metrics.chat_completion_sec = chat.completion;
    
    metrics.system_after = getSystemMetrics();
    
    fs.writeFileSync('after.json', JSON.stringify(metrics, null, 4));
    console.log("Benchmark complete.");
    console.log(JSON.stringify(metrics, null, 4));
}

run();
